sycl/include/sycl/ext/intel/esimd/detail/intrin.hpp

```diff
@@ -121,13 +121,13 @@ __esimd_rdindirect(__ESIMD_DNS::vector_type_t<T, N> Input,
 //
 template <typename T, int N, int M, int VStride, int Width, int Stride,
           int ParentWidth = 0>
-__ESIMD_INTRIN __ESIMD_DNS::vector_type_t<T, N>
+__ESIMD_INTRIN std::enable_if_t<M <= N, __ESIMD_DNS::vector_type_t<T, N>>
```

Comment:
1) The comment to this function is written as if was supposed to be a doxygen. It maybe converted to it by replacing '//' with '///'
2) We probably should also check 'assert(M % Width == 0);' from L108 in this SFINAE condition.
3) It may be useful to add restrictions to 'Stride' and 'VStride', e.g. 'Stride <= Width', 'VStride <= M/Width'. 
4) A stronger condition connecting VStride and Stride: 
((NumRows - 1)*VStride + (Width-1)*Stride < N)

(3) and (4) may make it not possible to write something like:
simd_val.select<1, HUGE_STRIDE_BIGGER_THAN_simd_val_size>, but such uses are useless anyways.

@sarnex @turinevgeny @fineg74 what do you think about adding more restrictions (3) and (4) here?

Reply:
Based on Gregory's comment is seems that OOB accesses on the GPU are not undefined and have some specified behavior.
In my experience this seems to be true, I have never seen a crash/hang because of OOB accesses on a GPU, only on the emulator.

If we add 3 and 4 from Slava's comment, we definitely may break existing programs that work on the GPU today. But from a user POV, I highly doubt any OOB accesses are intentional, so these restrictions would definitely have some value in finding bugs in user code.

If we are confident nobody would actually want to do an OOB access on a GPU intentionally, maybe we should try to do some internal testing and see if we break any tests. If we don't break any or only a few, we could proceed with the new restrictions, updating tests as neccessary. If it's a bloodbath, maybe we can reconsider.

But for restrictions that are specifically in the VC intrinsics spec for a specific intrinsic, we should definitely have restrictions for those.